Switch assets fast

Reizoukodesu opened the pull request:
Added a eyedropper tool toggle button for the editor which allows to switch quickly between tileset categories and select the clicked tile template as current brush. See #10643

Commit:
Added a WIP logic to switch between assets quickly

For now it only works for the tile assets.
To switch between the assets you just need to right click an
existing tile type of the desired tile type.

Commit:
Added eyedropper toggle button

pchote commented:
Could you please try and make this work as an `IEditorBrush` implementation, instead of adding the logic to the viewport widget?

Commit:
Refactored the editor eyedropper tool

Extracted the hardcoded code from the EditorViewportControllerWidget
Added mouseover effect while eyedropping

Commit:
Removed unecessary parenthesis

Commit:
Added newline between blocks

Commit:
Renamed wrong file name not matching class name

Commit:
Removed unused code

Sapena commented:
I've extracted the eyedropper brush logic to a new brush type.
I also added a visual feedback for the tool, I've called it EditorHighlightLayer. This could be used for other functions as well, as mentioned in [#10652](https://github.com/OpenRA/OpenRA/issues/10652#issuecomment-177339989). Should I extract the EditorHighlightLayer functionality to a different pull request?

Commit:
Fixed using directive orders

pchote commented on OpenRA.Mods.Common/Traits/World/EditorHighlightLayer.cs:
```diff
@@ -0,0 +1,81 @@
+		public virtual object Create(ActorInitializer init) { return new EditorHighlightLayer(init.Self, this); }
+	}
+
+	public class EditorHighlightLayer : IWorldLoaded, IPostRender
```
There's a lot of duplication between this and the existing `EditorSelectionLayer`. Can you merge the two together?

Sapena replied:
I couldn't find a good way to merge them together, so I extracted the shared logic to a parent class, is this ok?

pchote commented on OpenRA.Mods.Common/Widgets/EditorViewportControllerWidget.cs:
```diff
@@ -35,7 +36,28 @@ public EditorViewportControllerWidget(World world, WorldRenderer worldRenderer)
 			CurrentBrush = defaultBrush = new EditorDefaultBrush(this, worldRenderer);
 		}
 
+		public void ToggleEyedropping()
```
The modifications in this file aren't needed.  Just set the brush type directly from the `MapEditorLogic` (see how the copy/paste brush does it).

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Editor/TileSelectorLogic.cs:
```diff
@@ -50,6 +52,13 @@ public TileSelectorLogic(Widget widget, WorldRenderer worldRenderer, Ruleset mod
 
+		public void SwitchCategory(WorldRenderer worldRenderer, TileSet tileset, string category)
+		{
+			var tileCategorySelector = tileSelectorContainer.Get<DropDownButtonWidget>("TILE_CATEGORY");
```
This can be cached from the constructor.

pchote commented on mods/ra/chrome/editor.yaml:
```diff
@@ -394,3 +394,9 @@ Container@EDITOR_WORLD_ROOT:
+			X: 665
+			Width: 120
+			Height: 25
+			Text: Eyedropper tool
```
This label is too long, and overlaps the minimap.  I suggest renaming this to "Select terrain" and moving the button.

pchote commented on mods/ra/rules/world.yaml:
```diff
@@ -174,3 +174,4 @@ EditorWorld:
 	EditorActorLayer:
 	EditorResourceLayer:
 	EditorSelectionLayer:
+	EditorHighlightLayer:
```
Please add the feature to all the default mods, not just RA.

Commit:
Fixed multiple coding issues

Commit:
Accidentally added wrong file

Commit:
Fixed coding guideline issue

Sapena commented:
Hope all issues are fixed. Let me know if there's something more to do.

Mailaender commented:
Can you squash the fixup commits using `git rebase`?

pchote commented on OpenRA.Mods.Common/Traits/World/EditorCellRegionLayer.cs:
```diff
@@ -0,0 +1,60 @@
+
+		public virtual void RenderAfterWorld(WorldRenderer wr, Actor self)
+		{
+			if (wr.World.Type != WorldType.Editor)
```
You immediately return if this is false, too, so this check is a no-op.

pchote commented:
Any progress here @Sapena?

reaperrr commented:
There hasn't been any activity in two months, so I'm closing this for now to clear the PR queue.
@Sapena Feel free to reopen it when you have the time to rebase and adress the requests.
